runner/src/main.rs

```diff
@@ -595,19 +653,267 @@ struct Example {
+                    foreground: Box::new(backend_steps),
+                }
+            })
+    }
+}
+
+/// A construct representing either an Oak Classic or an Oak Functions application.  
+///
+/// The condition that only one of `classic` or `functions` should be non-empty is
+/// checked in each operation of this struct. If neither or both are empty, the
+/// operation panics with an error message.
 #[derive(serde::Deserialize, Debug)]
 #[serde(deny_unknown_fields)]
-struct Application {
+struct Applications {
+    classic: Option<ApplicationsList<ApplicationClassic>>,
+    functions: Option<ApplicationsList<ApplicationFunctions>>,
+}
+
+impl Applications {
+    fn is_classic(&self) -> bool {
+        match (&self.classic, &self.functions) {
+            (Some(_applications), None) => true,
+            (None, Some(_applications)) => false,
+            _ => panic!(INVALID_APP_COMBO_ERR),
+        }
+    }
+
+    fn empty_or_has_application_variant(&self, application_variant: &str) -> bool {
+        match (&self.classic, &self.functions) {
+            (Some(applications), None) => {
+                applications.empty_or_has_application_variant(application_variant)
+            }
+            (None, Some(applications)) => {
+                applications.empty_or_has_application_variant(application_variant)
+            }
+            _ => panic!(INVALID_APP_COMBO_ERR),
+        }
+    }
+
+    fn construct_application_build_steps(
+        &self,
+        application_variant: &str,
+        example_name: &str,
+    ) -> Vec<Step> {
+        match (&self.classic, &self.functions) {
+            (Some(applications), None) => {
+                if applications.is_empty() {
+                    vec![]
+                } else {
+                    let application = applications.get_application(application_variant);
+                    application.construct_application_build_steps(example_name)
+                }
+            }
+            (None, Some(applications)) => {
+                if applications.is_empty() {
+                    vec![]
+                } else {
+                    let application = applications.get_application(application_variant);
+                    application.construct_application_build_steps(example_name)
+                }
+            }
+            _ => panic!(INVALID_APP_COMBO_ERR),
+        }
+    }
+
+    fn construct_example_server_run_step(
+        &self,
+        opt: either::Either<&RunExamples, &RunFunctionsExamples>,
+        example: &Example,
+        run_clients: Step,
+    ) -> Step {
+        match (&self.classic, &self.functions) {
+            (Some(applications), None) => {
+                if applications.is_empty() {
+                    run_clients
+                } else {
+                    let opt = opt
+                        .left()
+                        .expect("Oak Classic application run with RunExamples options.");
```

Comment:
I cannot think of an actionable suggestion for this PR, but it feels there may be a type safe way of expressing these constraints instead of manually checking all the combinations. If / when we come up with one perhaps we can refactor this.

Reply:
I agree! I have been playing with a few ideas, without much luck. But I'll continue in a separate PR.